app/services/authentication/providers.rb

```diff
@@ -1,11 +1,70 @@
+# We require all authentication modules to make sure providers
+# are correctly preloaded both in development and in production and
+# ready to be used when needed at runtime
+Dir[Rails.root.join("app/services/authentication/**/*.rb")].each do |f|
+  require_dependency(f)
+end
+
 module Authentication
   module Providers
-    # TODO: [thepracticaldev/oss] raise exception if provider is available but not enabled for this app
-    # TODO: [thepracticaldev/oss] add available providers, enabled providers
+    # Retrieves a provider that is both available and enabled
     def self.get!(provider_name)
-      "Authentication::Providers::#{provider_name.titleize}".constantize
-    rescue NameError => e
-      raise ::Authentication::Errors::ProviderNotFound, e
+      name = provider_name.to_s.titleize
+
+      unless Authentication::Providers.const_defined?(name)
+        raise(
+          ::Authentication::Errors::ProviderNotFound,
+          "Provider #{name} is not available!",
+        )
+      end
+
+      unless enabled?(provider_name)
+        raise(
+          ::Authentication::Errors::ProviderNotEnabled,
+          "Provider #{name} is not enabled!",
+        )
+      end
+
+      Authentication::Providers.const_get(name)
+    end
+
+    # Returns available providers
+    def self.available
+      # the magic is done in <config/initializers/authentication_providers.rb>
+      Authentication::Providers::Provider.subclasses.map do |subclass|
+        subclass.name.demodulize.downcase.to_sym
+      end.sort
+    end
+
+    # Returns enabled providers
```

Comment:
Personally, I think all of these methods are named well enough that we don't need the additional comments.

Reply:
adding to the TODO as tests are too slow for only this change hahaha